### PR TITLE
[CARBONDATA-1926][CARBONDATA-1927][Pre-Aggregate] Expression support inside aggregate function for Query

### PR DESCRIPTION
**Pull request by kumarvishal09**
Support transforming of query plan for aggregate table when query aggregate function contains any expression
Support sub query  in Preaggregate table

**Note:**Need to add subquery test cases
 - [ ] Any interfaces changed?
 NA
 - [ ] Any backward compatibility impacted?
 NA
 - [ ] Document update required?
NA
 - [ ] Testing done
     Added UT to validate pre aggregate table selection and Data validation
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2364/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1152/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2578/

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/integration/spark/testsuite/preaggregate/TestPreAggregateExpressions.scala [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    checkAnswer(df, Seq(Row(2.0)))
+  }
+
+  test("test pre agg table selection with expression 5") {
+    val df = sql("select sum(case when age=27 then id else 0 end), sum(case when age=35 then id else 0 end) from maintable")
+    preAggTableValidator(df.queryExecution.analyzed, "maintable_agg4")
+    checkAnswer(df, Seq(Row(2.0,6.0)))
+  }
+
+  def preAggTableValidator(plan: LogicalPlan, actualTableName: String) : Unit ={
```
Comment (jackylk) [inline integration/spark-common-test/src/test/scala/org/apache/carbondata/integration/spark/testsuite/preaggregate/TestPreAggregateExpressions.scala]: add comment for this function

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/integration/spark/testsuite/preaggregate/TestPreAggregateExpressions.scala [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    preAggTableValidator(df.queryExecution.analyzed, "maintable_agg3")
+    checkAnswer(df, Seq(Row(2.0)))
+  }
+
+  test("test pre agg table selection with expression 5") {
+    val df = sql("select sum(case when age=27 then id else 0 end), sum(case when age=35 then id else 0 end) from maintable")
+    preAggTableValidator(df.queryExecution.analyzed, "maintable_agg4")
+    checkAnswer(df, Seq(Row(2.0,6.0)))
+  }
+
```
Comment (jackylk) [inline integration/spark-common-test/src/test/scala/org/apache/carbondata/integration/spark/testsuite/preaggregate/TestPreAggregateExpressions.scala]: add a testcase for subquery also

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala [hunk 3]
```diff
@@ -126,16 +127,17 @@ case class CarbonPreAggregateQueryRules(sparkSession: SparkSession) extends Rule
...
+          (carbonTable, logicalRelation)
 
         // below case for handling filter query
         // When plan has grouping expression, aggregate expression
         // filter expression
         case Aggregate(groupingExp,
-          aggregateExp,
-          Filter(filterExp,
-          CarbonSubqueryAlias(_, logicalRelation: LogicalRelation)))
+        aggregateExp,
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala]: unnecessary change

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2421/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1197/

**Comment by kumarvishal09** [comment]
Retest this please

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2612/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2426/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1202/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2625/

File: core/src/main/java/org/apache/carbondata/core/metadata/schema/table/AggregationDataMapSchema.java [hunk 4]
```diff
@@ -62,6 +62,12 @@
    */
   private int ordinal = Integer.MAX_VALUE;
 
+  /**
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/metadata/schema/table/AggregationDataMapSchema.java]: Remove unnecessary attributes like parentColumnToAggregationsMapping

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala [hunk 5]
```diff
@@ -330,6 +207,264 @@ case class CarbonPreAggregateQueryRules(sparkSession: SparkSession) extends Rule
...
+  }
+
+  /**
+   * Below method will be used to get aggregate expression
+   * @param logicalPlan
+   *               logical plan
+   * @return list of aggregate expression
+   */
+  def getAggregateExpFromChildDataMap(logicalPlan: LogicalPlan): Seq[AggregateExpression] = {
+    val list = scala.collection.mutable.HashSet.empty[AggregateExpression]
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala]: Should use List or LinkedHasSet as the order of insertion is required for mapping

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala [hunk 6]
```diff
@@ -330,6 +207,264 @@ case class CarbonPreAggregateQueryRules(sparkSession: SparkSession) extends Rule
...
+   *                        logical relation
+   * @return logical plan
+   */
+  def getLogicalPlanFromAggExp(aggExp: AggregateExpression,
+      tableName: String,
+      databaseName: String,
+      logicalRelation: LogicalRelation): LogicalPlan = {
+    // adding the preAGG UDF, so pre aggregate data loading rule and query rule will not
+    // be applied
+    val query = new CarbonSpark2SqlParser()
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala]: Don't create parser every time, pass from caller

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala [hunk 7]
```diff
@@ -330,6 +207,264 @@ case class CarbonPreAggregateQueryRules(sparkSession: SparkSession) extends Rule
...
+   * @return if plan is valid for tranformation, parent table, parent logical relaion
+   */
+  def validatePlanAndGetFields(logicalPlan: LogicalPlan,
+      list: scala.collection.mutable.HashSet[QueryColumn],
+      qAggExprs: scala.collection.mutable.HashSet[AggregateExpression]): (Boolean,
+    CarbonTable, LogicalRelation) = {
+    var isValidPlan = false
+    var pTable: CarbonTable = null
+    var qLRelation: LogicalRelation = null
+    logicalPlan.transform {
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala]: It seems Join condition of two plans is not handled here. Please handle Join, Union cases also using recursion.

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala [hunk 8]
```diff
@@ -282,6 +136,7 @@ case class CarbonPreAggregateQueryRules(sparkSession: SparkSession) extends Rule
         val listFilterColumn = list
           .filter(queryColumn => queryColumn.getAggFunction.isEmpty && queryColumn.isFilterColumn)
           .toList
+        val isProjectionColumnPresent = (listProjectionColumn.size + listFilterColumn.size) > 0
         // getting all the aggregation columns
         val listAggregationColumn = list.filter(queryColumn => !queryColumn.getAggFunction.isEmpty)
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala]: This code is unused now , please remove all code related to it.

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala [hunk 9]
```diff
@@ -1124,92 +1328,57 @@ case class CarbonPreAggregateQueryRules(sparkSession: SparkSession) extends Rule
...
+   * @param aggExp
+   * aggregate expression
    * @return list of fields
    */
   def validateAggregateFunctionAndGetFields(carbonTable: CarbonTable,
-      aggFunctions: AggregateFunction,
-      tableName: String
-  ): Seq[QueryColumn] = {
+      aggExp: AggregateExpression): Seq[AggregateExpression] = {
     val changedDataType = true
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala]: Remove unused variable

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala [hunk 10]
```diff
@@ -933,121 +1143,122 @@ case class CarbonPreAggregateQueryRules(sparkSession: SparkSession) extends Rule
...
   def getUpdatedAggregateExpressionForChild(aggExp: AggregateExpression,
-      dataMapSchema: DataMapSchema,
-      attributes: Seq[AttributeReference]):
+      dataMapSchema: AggregationDataMapSchema,
+      attributes: Seq[AttributeReference],
+      expLogicalPlanToColumnSchemaMapping: Option[Map[LogicalPlan, ColumnSchema]],
+      parentTable: CarbonTable,
+      logicalRelation: LogicalRelation):
   Expression = {
+    val updatedAggExp = getUpdateAggregateExpressions(aggExp)
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala]: Add code comment here

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala [hunk 11]
```diff
@@ -294,12 +149,30 @@ case class CarbonPreAggregateQueryRules(sparkSession: SparkSession) extends Rule
...
+          selectedDataMapSchemas
+        } else {
+          carbonTable.getTableInfo.getDataMapSchemaList
+        }
+        val aggExpLogicalPlans = aggregateExpressions.map { queryAggExp =>
+          getLogicalPlanFromAggExp(queryAggExp,
+            carbonTable.getTableName,
+            carbonTable.getDatabaseName, logicalRelation)
+        }.toSeq
+        if(aggregateExpressions.size > 0 && selectedAggMaps.size > 0) {
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala]: add comments

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1219/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2451/

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/integration/spark/testsuite/preaggregate/TestPreAggregateExpressions.scala (hunk 2 above)
Comment (kumarvishal09) [reply integration/spark-common-test/src/test/scala/org/apache/carbondata/integration/spark/testsuite/preaggregate/TestPreAggregateExpressions.scala]: add TestPreAggregateWithSubQuery

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2508/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1283/

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala [hunk 12]
```diff
@@ -18,28 +18,46 @@
...
 
+/**
+ * model class to store aggregate expression logical plan
+ * and its column schema mapping
+ * @param logicalPlan
+ * logical plan of aggregate expression
+ * @param columnSchema
+ * column schema from table
+ */
+case class AggExpToColumnMappingModel(var logicalPlan: LogicalPlan,
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala]: Move `var logicalPlan: LogicalPlan` to next line

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala [hunk 13]
```diff
@@ -98,238 +143,499 @@ case class CarbonPreAggregateQueryRules(sparkSession: SparkSession) extends Rule
...
-            isValidPlan = extractQueryColumnFromFilterExp(filterExp, list, carbonTable, tableName)
+  /**
+   * Below method will be used to update the sort expression
+   * @param sortExp
+   * sort order expression in query
+   * @return updated sort expression
+   */
+  def updateSortExpression(sortExp : Seq[SortOrder]) : Seq[SortOrder] = {
+     sortExp map { order =>
+      order.child match {
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala]: Better do the transform instead of match as in case if Alias comes match cannot work

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala [hunk 14]
```diff
@@ -98,238 +143,499 @@ case class CarbonPreAggregateQueryRules(sparkSession: SparkSession) extends Rule
...
+        case attr: AttributeReference =>
+          val childExp = updatedExpression.find { p => p._1.sameRef(attr) }
+          if (childExp.isDefined) {
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala]: Better use `match { case }` instead of if else

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala [hunk 15]
```diff
@@ -98,238 +143,499 @@ case class CarbonPreAggregateQueryRules(sparkSession: SparkSession) extends Rule
...
+            val newExpression = AttributeReference(
+              childExp.get._2.name,
+              childExp.get._2.dataType,
+              childExp.get._2.nullable,
+              childExp.get._2.metadata)(childExp.get._2.exprId, attr.qualifier, attr.isGenerated)
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala]: why you use `attr.qualifier `  intead of `childExp.qualifier`

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala [hunk 16]
```diff
@@ -98,238 +143,499 @@ case class CarbonPreAggregateQueryRules(sparkSession: SparkSession) extends Rule
...
+   * @param expressions
+   * sequence of expression like group by
+   * @return updated expressions
+   */
+  def updateExpression(expressions : Seq[Expression]) : Seq[Expression] = {
+    val newExp = expressions map { expression =>
+      expression transform {
+        case attr: AttributeReference =>
+          val childExp = updatedExpression.find { p => p._1.sameRef(attr) }
+          if (childExp.isDefined) {
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala]: Better use match { case } instead of if else

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala [hunk 17]
```diff
@@ -98,238 +143,499 @@ case class CarbonPreAggregateQueryRules(sparkSession: SparkSession) extends Rule
...
+            val newExpression = AttributeReference(
+              childExp.get._2.name,
+              childExp.get._2.dataType,
+              childExp.get._2.nullable,
+              childExp.get._2.metadata)(childExp.get._2.exprId, attr.qualifier, attr.isGenerated)
+            newExpression
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala]: not required to assign to variable

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala [hunk 18]
```diff
@@ -98,238 +143,499 @@ case class CarbonPreAggregateQueryRules(sparkSession: SparkSession) extends Rule
...
+
+  /**
+   * Below method will be used to updated the named expression like aggregate expression
+   * @param namedExpression
+   * any named expression like aggregate expression
+   * @return updated named expression
+   */
+  def updateNamedExpression(namedExpression: Seq[NamedExpression]) : Seq[NamedExpression] = {
+    namedExpression map {
+      case attr: AttributeReference =>
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala]: use transform inside map , then you no need to handle Alias separetly.

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala [hunk 19]
```diff
@@ -98,238 +143,499 @@ case class CarbonPreAggregateQueryRules(sparkSession: SparkSession) extends Rule
...
+    newExp
+  }
+
+  /**
+   * Below method will be used to updated the named expression like aggregate expression
+   * @param namedExpression
+   * any named expression like aggregate expression
+   * @return updated named expression
+   */
+  def updateNamedExpression(namedExpression: Seq[NamedExpression]) : Seq[NamedExpression] = {
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala]: You can directly use `updateExpression` no need of this method. `NamedExpression` is derived from `Expression`. So after updation you can type cast.

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala [hunk 20]
```diff
@@ -98,238 +143,499 @@ case class CarbonPreAggregateQueryRules(sparkSession: SparkSession) extends Rule
...
+    }
+  }
+
+  /**
+   * Below method will be used to updated condition expression
+   * @param conditionExp
+   * any condition expression join condition or filter condition
+   * @return updated condition expression
+   */
+  def updateConditionExpression(conditionExp: Option[Expression]): Option[Expression] = {
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala]: Here also you use updateExpression method, no need of this method

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala [hunk 21]
```diff
@@ -98,238 +143,499 @@ case class CarbonPreAggregateQueryRules(sparkSession: SparkSession) extends Rule
...
+   *    5.2 timeseries function is valid function or not
+   *
+   * @param logicalPlan
+   * parent logical plan
+   * @return transformed plan
+   */
+  def transformPreAggQueryPlan(logicalPlan: LogicalPlan): LogicalPlan = {
+    val updatedPlan = logicalPlan.transform {
+      // case for aggregation query
+      case agg@Aggregate(grExp,
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala]: MOve down `grExp` to next line

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala [hunk 22]
```diff
@@ -98,238 +143,499 @@ case class CarbonPreAggregateQueryRules(sparkSession: SparkSession) extends Rule
...
+      aggExp,
+      child@CarbonSubqueryAlias(_, logicalRelation: LogicalRelation))
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala]: Indentation is wrong

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala (hunk 15 above)
Comment (kumarvishal09) [reply integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala]: childExp.qualifier will not have table alias name in case of join we need to table alias name

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala [hunk 23]
```diff
@@ -806,67 +824,105 @@ case class CarbonPreAggregateQueryRules(sparkSession: SparkSession) extends Rule
     val updatedAggExp = aggregateExpressions.map {
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala]: It seems lot of duplicate code, better do transform and use single  `case attr: AttributeReference` handle all here

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala [hunk 24]
```diff
@@ -98,238 +143,499 @@ case class CarbonPreAggregateQueryRules(sparkSession: SparkSession) extends Rule
...
-             logicalRelation.relation.asInstanceOf[CarbonDatasourceHadoopRelation].carbonRelation.
-               metaData.hasAggregateDataMapSchema =>
-          val (carbonTable, tableName) = getCarbonTableAndTableName(logicalRelation)
-          isValidPlan = extractQueryColumnsFromAggExpression(groupingExp,
-            aggregateExp,
+        } else {
+          agg
+        }
+      // case of handling aggregation query with filter
+      case agg@Aggregate(grExp,
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala]: How about with Sort ?

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala [hunk 25]
```diff
@@ -98,238 +143,499 @@ case class CarbonPreAggregateQueryRules(sparkSession: SparkSession) extends Rule
...
+      carbonTable: CarbonTable,
+      logicalRelation: LogicalRelation,
+      queryAggExpLogicalPlans: Seq[LogicalPlan]): Seq[DataMapSchema] = {
+    def validateDataMap(dataMap: DataMapSchema,
+        aggExpLogicalPlans: Seq[LogicalPlan]): Boolean = {
+      val mappingModel = getLogicalPlanForAggregateExpression(dataMap,
+        carbonTable,
+        logicalRelation)
+      aggExpLogicalPlans.forall{
+        p => mappingModel.exists(m => p.sameResult(m.logicalPlan))
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala]: move p => to up

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2674/

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala [hunk 26]
```diff
@@ -1218,112 +1180,125 @@ case class CarbonPreAggregateQueryRules(sparkSession: SparkSession) extends Rule
...
-    }
-
   /**
    * Called by PreAggregateLoadingRules to validate if plan is valid for applying rules or not.
    * If the plan has PreAggLoad i.e Loading UDF and does not have PreAgg i.e Query UDF then it is
    * valid.
-   *
    * @param namedExpression
-   * @return
+   * named expressions
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala]: move it up. comment should like
`@param namedExpression named expressions`

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2551/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1326/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2562/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2722/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2566/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2568/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2732/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2733/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2734/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2735/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2738/

**Comment by kumarvishal09** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1338/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2575/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2577/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1343/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2585/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1349/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2594/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2755/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1360/

**Comment by ravipesala** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2612/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1377/

**Comment by sraghunandan** [comment]
retest sdv please

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2787/

**Comment by ravipesala** [comment]
LGTM
